core/src/main/java/jenkins/model/Jenkins.java

```diff
@@ -3285,13 +3287,31 @@ public void run(Reactor session) throws Exception {
      * Save the settings to a file.
      */
     public synchronized void save() throws IOException {
-        if(BulkChange.contains(this))   return;
-
-        if (initLevel == InitMilestone.COMPLETED) {
+        InitMilestone currentMilestone = initLevel;
+
+        if (!configLoaded) {
+            // someone is trying to save the config before all extensions are loaded (and possibly after as the task
+            // may run in parallel with other tasks.  OMG...!!! this is generally very bad and can lead to dataloss
+            LOGGER.log(Level.SEVERE,
```

Comment:
Seems unnecessary to both log a stack trace _and_ throw an exception. My understanding of this situation is that we expect data loss, so something is really broken and we should not return normally from `save()`. Therefore delete the `log` call and expand the message in the `IOException`.

(Arguably we should be throwing up an `IllegalStateException`: an `IOException` is used to indicate that a save was attempted and failed due to filesystem problems, which is normally caught and logged, with the expectation that a subsequent save would persist the same data anyway, so at worst we are losing a bit of robustness against a JVM crash. This case is quite different.)

Reply:
I am not convinced that all callers of save will appropriately log an `IOException` coming from `save` rather than just something like
`LOGGER.log(Level.FINE, "could not save config")`
Given this has such a potential for damage I wanted to ensure that
1) the admin has a clearly visible log which shows something bad has happened along with a stack trace to debug
2) the caller at least knows something has gone wrong.

Because I do not trust all the plugins in the eco system I think just throwing an `Exception` here would be a bad idea.

I originally thought about throwing an `IllegalStateException` but decided against it as the load will happen and Jenkins will be at least somewhat functional whereas if we kill a calling thread then this can make something bad even worse.  But I can be convinced otherwise.

Possibly this should create a new [`AdministrativeError`](https://javadoc.jenkins.io/index.html?hudson/util/AdministrativeError.html) for the allerting as well.   Ideally I would throw a [`BootException`](https://javadoc.jenkins.io/index.html?hudson/util/BootFailure.html) but the stack may not be Jenkins and this may not cause the correct thing to do.

Another possibility is to set a flag and then check this after loading and if the flag is set then throw the `BootFailure`.

comments on any proposals welcome.

Reply:
IMO outcome of this change should be: 1) avoid data loss and 2) help identify the plugin/code at fault.

Current state achieve that (log and throw), however it could lead to double logging if the caller is also logging the exception. Using `IllegalStateException` also gives us what we want but will kill the caller thread, as James said. We don't know if that thread is a _vital_ one or not, so I think I'd stay in the safe side and assume the double logging (which is annoying but not critical).

Reply:
I am leaning more towards `IllegalStateException`.  yes we may kill a caller thread, but at the same time that's better than wiping configuration that the same thread would have done.

Calling threads could be killed for all manner of issues (OutOfMemory, if they have an `ExceptionHandler` then this can be used to reincarnate the thread, otherwise we have the top level `ExceptionHandler` that will (along with the log above) say something terrible has happened.

Kind of like shooting the burgler rather than politely asking them to leave :)